File: ios/Classes/KidozAdsPlugin.swift
import Flutter
import KidozSDK
import UIKit

/// This package's own error codes. Kidoz has none of its own to pass through.
///
/// `KidozError` exposes a description and no status enum — Kidoz's own sample
/// app reads nothing else from it. So the distinction a caller actually needs
/// (is this an empty auction, or did something break?) has to be made here,
/// from *which delegate callback fired*, because that is the only place the
/// information exists. The names match the Android side exactly, so
/// `KidozAdError.code` means one thing in Dart.
enum KidozErrorCode {
    /// A load callback failed. The signal to fall through to another network.
    static let noFill = "NO_FILL"
    /// The ad loaded and then failed to present.
    static let showFailed = "SHOW_FAILED"
    /// A load was attempted before initialization completed.
    static let notInitialized = "NOT_INITIALIZED"
    /// This plugin's own faults, and anything unclassified.
    static let internalError = "INTERNAL_ERROR"
}

/// Flattens a Kidoz failure into the shape `KidozAdError.fromMap` expects.
func kidozEventMap(_ error: KidozError?, code: String) -> [String: Any?] {
    return [
        "code": code,
        "message": error?.description ?? "No message provided",
    ]
}

/// Bridges the Kidoz iOS SDK onto one method channel.
///
/// The iOS SDK's load entry points are **static** and hand the ad to the
/// delegate rather than returning it — `KidozInterstitialAd.load(delegate:)` —
/// so there is no object to key on until the load callback fires. Ads are
/// stashed in `fullScreenAds` against the id Dart minted *inside* that
/// callback, and each delegate closes over its id to do it.
///
/// `delegates` is a second dictionary rather than a property on the ad for the
/// usual UIKit reason: Kidoz holds its delegate weakly, so letting one go out
/// of scope means the ad loads and no callback ever arrives.
public class KidozAdsPlugin: NSObject, FlutterPlugin {

    private let channel: FlutterMethodChannel
    private var fullScreenAds: [Int: AnyObject] = [:]
    private var delegates: [Int: AnyObject] = [:]

    init(channel: FlutterMethodChannel) {
        self.channel = channel
        super.init()
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "kidoz_ads",
            binaryMessenger: registrar.messenger()
        )
        let instance = KidozAdsPlugin(channel: channel)
        registrar.addMethodCallDelegate(instance, channel: channel)

        registrar.register(
            KidozBannerViewFactory(messenger: registrar.messenger()) {
                [weak instance] adId, event, arguments in
                instance?.sendEvent(adId: adId, event: event, arguments: arguments)
            },
            withId: "kidoz_ads/banner"
        )
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = call.arguments as? [String: Any] ?? [:]

        switch call.method {
        case "initialize":
            initialize(arguments, result)
        case "loadFullScreenAd":
            loadFullScreenAd(arguments, result)
        case "showFullScreenAd":
            showFullScreenAd(arguments, result)
        case "disposeAd":
            if let adId = arguments["adId"] as? Int {
                fullScreenAds.removeValue(forKey: adId)
                delegates.removeValue(forKey: adId)
            }
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func initialize(_ arguments: [String: Any], _ result: @escaping FlutterResult) {
        guard
            let publisherId = arguments["publisherId"] as? String, !publisherId.isEmpty,
            let securityToken = arguments["securityToken"] as? String, !securityToken.isEmpty
        else {
            result(FlutterError(
                code: "INVALID_CREDENTIALS",
                message: "publisherId and securityToken must not be empty",
                details: nil
            ))
            return
        }

        let delegate = InitDelegate(result: result)
        // Held for the life of the process: Kidoz's init delegate is weak like
        // the rest, and initialization is one-way so there is nothing to
        // release it on.
        initDelegate = delegate

        Kidoz.instance.initialize(
            publisherID: publisherId,
            securityToken: securityToken,
            delegate: delegate
        )
    }

    private var initDelegate: InitDelegate?

    private func loadFullScreenAd(_ arguments: [String: Any], _ result: @escaping FlutterResult) {
        guard let adId = arguments["adId"] as? Int else {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "adId is required",
                details: nil
            ))
            return
        }

        // Checked here as well as in Dart: the Dart guard catches the ordinary
        // programming error, but a caller that raced initialize() would reach
        // a static load() that returns without ever calling back, and the
        // caller's chain would stall rather than falling through.
        guard Kidoz.instance.isSDKInitialized() else {
            sendEvent(
                adId: adId,
                event: "loadFailed",
                arguments: [
                    "code": KidozErrorCode.notInitialized,
                    "message": "Kidoz SDK is not initialized",
                ]
            )
            result(nil)
            return
        }

        let rewarded = arguments["rewarded"] as? Bool ?? false
        let store: (Int, AnyObject) -> Void = { [weak self] adId, ad in
            self?.fullScreenAds[adId] = ad
        }
        let forget: (Int) -> Void = { [weak self] adId in
            self?.fullScreenAds.removeValue(forKey: adId)
        }
        let send: (Int, String, [String: Any?]) -> Void = { [weak self] adId, event, arguments in
            self?.sendEvent(adId: adId, event: event, arguments: arguments)
        }

        if rewarded {
            let delegate = RewardedAdDelegate(
                adId: adId, store: store, forget: forget, sendEvent: send
            )
            delegates[adId] = delegate
            KidozRewardedAd.load(delegate: delegate)
        } else {
            let delegate = InterstitialAdDelegate(
                adId: adId, store: store, forget: forget, sendEvent: send
            )
            delegates[adId] = delegate
            KidozInterstitialAd.load(delegate: delegate)
        }
        result(nil)
    }

    private func showFullScreenAd(_ arguments: [String: Any], _ result: @escaping FlutterResult) {
        guard let adId = arguments["adId"] as? Int else {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "adId is required",
                details: nil
            ))
            return
        }

        guard let ad = fullScreenAds[adId] else {
            sendEvent(
                adId: adId,
                event: "showFailed",
                arguments: [
                    "code": KidozErrorCode.internalError,
                    "message": "Ad \(adId) is gone",
                ]
            )
            result(nil)
            return
        }

        guard let presenter = Self.topViewController() else {
            sendEvent(
                adId: adId,
                event: "showFailed",
                arguments: [
                    "code": KidozErrorCode.internalError,
                    "message": "No view controller available to present from",
                ]
            )
            result(nil)
            return
        }

        // isLoaded() is checked because show() on a spent or unloaded Kidoz ad
        // reports nothing at all — no delegate callback — which would leave the
        // caller's chain waiting on its own timeout.
        switch ad {
        case let interstitial as KidozInterstitialAd where interstitial.isLoaded():
            interstitial.show(viewController: presenter)
        case let rewarded as KidozRewardedAd where rewarded.isLoaded():
            rewarded.show(viewController: presenter)
        default:
            sendEvent(
                adId: adId,
                event: "showFailed",
                arguments: [
                    "code": KidozErrorCode.internalError,
                    "message": "Ad \(adId) is not ready",
                ]
            )
        }
        result(nil)
    }

    fileprivate func sendEvent(adId: Int, event: String, arguments: [String: Any?]) {
        var payload = arguments
        payload["adId"] = adId
        payload["event"] = event
        DispatchQueue.main.async { [weak self] in
            self?.channel.invokeMethod("onAdEvent", arguments: payload)
        }
    }

    /// The frontmost view controller, walking past anything already presented.
    ///
    /// Presenting from a controller that is itself covered throws a UIKit
    /// warning and shows nothing, which is easy to hit when an ad is triggered
    /// from a Flutter route shown over a native modal.
    static func topViewController() -> UIViewController? {
        let root = UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap { $0.windows }
            .first { $0.isKeyWindow }?
            .rootViewController

        var top = root
        while let presented = top?.presentedViewController {
            top = presented
        }
        return top
    }
}

/// Reports the one-shot outcome of `Kidoz.instance.initialize` back to Dart.
final class InitDelegate: NSObject, KidozInitDelegate {

    private var result: FlutterResult?

    init(result: @escaping FlutterResult) {
        self.result = result
        super.init()
    }

    func onInitSuccess() {
        // Nulled after firing: a FlutterResult must be called exactly once, and
        // an SDK that reported both success and failure would otherwise crash
        // the engine rather than the ad.
        result?(nil)
        result = nil
    }

    func onInitError(_ error: String) {
        result?(FlutterError(code: "INIT_FAILED", message: error, details: nil))
        result = nil
    }
}

/// One interstitial's delegate, tagged with the id its events belong to.
final class InterstitialAdDelegate: NSObject, KidozInterstitialDelegate {

    private let adId: Int
    private let store: (Int, AnyObject) -> Void
    private let forget: (Int) -> Void
    private let sendEvent: (Int, String, [String: Any?]) -> Void

    init(
        adId: Int,
        store: @escaping (Int, AnyObject) -> Void,
        forget: @escaping (Int) -> Void,
        sendEvent: @escaping (Int, String, [String: Any?]) -> Void
    ) {
        self.adId = adId
        self.store = store
        self.forget = forget
        self.sendEvent = sendEvent
        super.init()
    }

    func onInterstitialAdLoaded(kidozInterstitialAd: KidozInterstitialAd) {
        store(adId, kidozInterstitialAd)
        sendEvent(adId, "loaded", [:])
    }

    func onInterstitialAdFailedToLoad(kidozError: KidozError) {
        sendEvent(adId, "loadFailed", kidozEventMap(kidozError, code: KidozErrorCode.noFill))
    }

    func onInterstitialAdShown(kidozInterstitialAd: KidozInterstitialAd) {
        sendEvent(adId, "shown", [:])
    }

    func onInterstitialAdFailedToShow(
        kidozInterstitialAd: KidozInterstitialAd,
        kidozError: KidozError
    ) {
        forget(adId)
        sendEvent(adId, "showFailed", kidozEventMap(kidozError, code: KidozErrorCode.showFailed))
    }

    func onInterstitialImpression(kidozInterstitialAd: KidozInterstitialAd) {
        sendEvent(adId, "impression", [:])
    }

    func onInterstitialAdClosed(kidozInterstitialAd: KidozInterstitialAd) {
        forget(adId)
        sendEvent(adId, "closed", [:])
    }
}

/// One rewarded ad's delegate, tagged with the id its events belong to.
final class RewardedAdDelegate: NSObject, KidozRewardedDelegate {

    private let adId: Int
    private let store: (Int, AnyObject) -> Void
    private let forget: (Int) -> Void
    private let sendEvent: (Int, String, [String: Any?]) -> Void

    init(
        adId: Int,
        store: @escaping (Int, AnyObject) -> Void,
        forget: @escaping (Int) -> Void,
        sendEvent: @escaping (Int, String, [String: Any?]) -> Void
    ) {
        self.adId = adId
        self.store = store
        self.forget = forget
        self.sendEvent = sendEvent
        super.init()
    }

    func onRewardedAdLoaded(kidozRewardedAd: KidozRewardedAd) {
        store(adId, kidozRewardedAd)
        sendEvent(adId, "loaded", [:])
    }

    func onRewardedAdFailedToLoad(kidozError: KidozError) {
        sendEvent(adId, "loadFailed", kidozEventMap(kidozError, code: KidozErrorCode.noFill))
    }

    func onRewardedAdShown(kidozRewardedAd: KidozRewardedAd) {
        sendEvent(adId, "shown", [:])
    }

    func onRewardedAdFailedToShow(kidozRewardedAd: KidozRewardedAd, kidozError: KidozError) {
        forget(adId)
        sendEvent(adId, "showFailed", kidozEventMap(kidozError, code: KidozErrorCode.showFailed))
    }

    /// Fires before `onRewardedAdClosed`, which is what lets a caller treat
    /// closure as terminal and still know whether a reward was earned.
    ///
    /// Kidoz passes no reward payload — no name, no amount — because there is
    /// no dashboard field to configure one. The reward is the signal.
    func onRewardReceived(kidozRewardedAd: KidozRewardedAd) {
        sendEvent(adId, "rewardReceived", [:])
    }

    func onRewardedImpression(kidozRewardedAd: KidozRewardedAd) {
        sendEvent(adId, "impression", [:])
    }

    func onRewardedAdClosed(kidozRewardedAd: KidozRewardedAd) {
        forget(adId)
        sendEvent(adId, "closed", [:])
    }
}

File: android/src/main/kotlin/com/illuminationdevelopment/kidoz_ads/KidozAdsPlugin.kt
package com.illuminationdevelopment.kidoz_ads

import android.app.Activity
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import net.kidoz.ads.fullscreen.interstial.KidozInterstitialAd
import net.kidoz.ads.fullscreen.interstial.KidozInterstitialAdCallback
import net.kidoz.ads.fullscreen.rewarded.KidozRewardedAd
import net.kidoz.ads.fullscreen.rewarded.KidozRewardedAdCallback
import net.kidoz.sdk.Kidoz
import net.kidoz.sdk.KidozError
import net.kidoz.sdk.KidozInitializationListener

/**
 * Bridges the Kidoz Android SDK onto one method channel.
 *
 * Three constraints from the SDK shape everything here:
 *
 *  - **`Kidoz.initialize` takes an Activity, not an application Context**, and
 *    the SDK's own docs say to call it from `MainActivity.onCreate`. A Flutter
 *    plugin attaches to the engine before it attaches to an activity, so
 *    `initialize` has to fail loudly when no activity is bound rather than
 *    quietly passing the application context and getting a WebView that cannot
 *    present.
 *  - **The load entry points are static and hand the ad to the callback**
 *    (`KidozInterstitialAd.load(activity, callback)`), so there is no object to
 *    key on until `onAdLoaded` fires. Ads are therefore stashed in
 *    [fullScreenAds] against the id Dart minted *inside* that callback, and the
 *    callback closes over the id to do it.
 *  - Kidoz's callbacks are not documented as main-thread-only, but its views
 *    are WebViews. Every entry point hops to the main looper rather than
 *    trusting the caller, because Flutter guarantees method calls arrive on the
 *    platform thread and says nothing about which thread that is.
 */
class KidozAdsPlugin : FlutterPlugin, ActivityAware, MethodChannel.MethodCallHandler {

    private lateinit var channel: MethodChannel
    private var activity: Activity? = null

    /**
     * Loaded ads, by the id Dart minted. The value is `Any` because Kidoz's two
     * full-screen types share no supertype worth naming — `show()` is declared
     * separately on each — so the cast happens at the one place that shows.
     */
    private val fullScreenAds = mutableMapOf<Int, Any>()
    private val mainHandler = Handler(Looper.getMainLooper())

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)

        binding.platformViewRegistry.registerViewFactory(
            BANNER_VIEW_TYPE,
            KidozBannerViewFactory({ activity }) { adId, event, arguments ->
                sendEvent(adId, event, arguments)
            },
        )
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        mainHandler.post { fullScreenAds.clear() }
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initialize" -> initialize(call, result)
            "loadFullScreenAd" -> loadFullScreenAd(call, result)
            "showFullScreenAd" -> showFullScreenAd(call, result)
            "disposeAd" -> {
                val adId = call.argument<Int>("adId")
                mainHandler.post { adId?.let(fullScreenAds::remove) }
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    private fun initialize(call: MethodCall, result: MethodChannel.Result) {
        val publisherId = call.argument<String>("publisherId")
        val securityToken = call.argument<String>("securityToken")

        if (publisherId.isNullOrBlank() || securityToken.isNullOrBlank()) {
            result.error(
                "INVALID_CREDENTIALS",
                "publisherId and securityToken must not be empty",
                null,
            )
            return
        }

        // Kidoz needs a foreground Activity to initialize. Reporting it here
        // means the caller learns while it can still fall through to another
        // network, rather than at the first load.
        val activity = this.activity ?: run {
            result.error(
                "NO_ACTIVITY",
                "Kidoz.initialize needs a foreground Activity; none is attached",
                null,
            )
            return
        }

        mainHandler.post {
            Kidoz.initialize(
                activity,
                publisherId,
                securityToken,
                object : KidozInitializationListener {
                    override fun onInitSuccess() {
                        result.success(null)
                    }

                    override fun onInitError(error: KidozError) {
                        result.error(
                            "INIT_FAILED",
                            error.message ?: "Kidoz SDK failed to initialize",
                            null,
                        )
                    }
                },
            )
        }
    }

    private fun loadFullScreenAd(call: MethodCall, result: MethodChannel.Result) {
        val adId = call.argument<Int>("adId")
        val rewarded = call.argument<Boolean>("rewarded") ?: false

        if (adId == null) {
            result.error("INVALID_ARGUMENTS", "adId is required", null)
            return
        }

        val activity = this.activity ?: run {
            result.error(
                "NO_ACTIVITY",
                "Kidoz full-screen ads need a foreground Activity; none is attached",
                null,
            )
            return
        }

        mainHandler.post {
            if (rewarded) {
                KidozRewardedAd.load(activity, RewardedCallback(adId))
            } else {
                KidozInterstitialAd.load(activity, InterstitialCallback(adId))
            }
        }
        result.success(null)
    }

    private fun showFullScreenAd(call: MethodCall, result: MethodChannel.Result) {
        val adId = call.argument<Int>("adId")
        if (adId == null) {
            result.error("INVALID_ARGUMENTS", "adId is required", null)
            return
        }

        mainHandler.post {
            when (val ad = fullScreenAds[adId]) {
                null -> sendEvent(
                    adId,
                    "showFailed",
                    mapOf(
                        "code" to KidozErrorCode.INTERNAL_ERROR,
                        "message" to "Ad $adId is gone",
                    ),
                )
                is KidozInterstitialAd -> ad.show()
                is KidozRewardedAd -> ad.show()
                else -> sendEvent(
                    adId,
                    "showFailed",
                    mapOf(
                        "code" to KidozErrorCode.INTERNAL_ERROR,
                        "message" to "Ad $adId has an unexpected type",
                    ),
                )
            }
        }
        result.success(null)
    }

    /** Bridges one interstitial's callbacks onto [sendEvent]. */
    private inner class InterstitialCallback(private val adId: Int) :
        KidozInterstitialAdCallback {

        override fun onAdLoaded(ad: KidozInterstitialAd) {
            fullScreenAds[adId] = ad
            sendEvent(adId, "loaded")
        }

        override fun onAdFailedToLoad(error: KidozError) =
            sendEvent(adId, "loadFailed", error.toEventMap(KidozErrorCode.NO_FILL))

        override fun onAdShown(ad: KidozInterstitialAd) = sendEvent(adId, "shown")

        override fun onAdFailedToShow(ad: KidozInterstitialAd, error: KidozError) {
            fullScreenAds.remove(adId)
            sendEvent(adId, "showFailed", error.toEventMap(KidozErrorCode.SHOW_FAILED))
        }

        override fun onAdImpression(ad: KidozInterstitialAd) = sendEvent(adId, "impression")

        override fun onAdClosed(ad: KidozInterstitialAd) {
            // Kidoz's own sample nulls its reference here, and the SDK requires
            // a fresh load to show again. Dropping it now means a stray show()
            // reports "Ad is gone" rather than failing silently inside the SDK.
            fullScreenAds.remove(adId)
            sendEvent(adId, "closed")
        }
    }

    /** Bridges one rewarded ad's callbacks onto [sendEvent]. */
    private inner class RewardedCallback(private val adId: Int) : KidozRewardedAdCallback {

        override fun onAdLoaded(ad: KidozRewardedAd) {
            fullScreenAds[adId] = ad
            sendEvent(adId, "loaded")
        }

        override fun onAdFailedToLoad(error: KidozError) =
            sendEvent(adId, "loadFailed", error.toEventMap(KidozErrorCode.NO_FILL))

        override fun onAdShown(ad: KidozRewardedAd) = sendEvent(adId, "shown")

        override fun onAdFailedToShow(ad: KidozRewardedAd, error: KidozError) {
            fullScreenAds.remove(adId)
            sendEvent(adId, "showFailed", error.toEventMap(KidozErrorCode.SHOW_FAILED))
        }

        override fun onAdImpression(ad: KidozRewardedAd) = sendEvent(adId, "impression")

        /**
         * Fires before [onAdClosed], which is what lets a caller treat closure
         * as terminal and still know whether a reward was earned.
         *
         * Kidoz passes no reward payload — no name, no amount — because there
         * is no dashboard field to configure one. The reward is the signal.
         */
        override fun onRewardReceived(ad: KidozRewardedAd) = sendEvent(adId, "rewardReceived")

        override fun onAdClosed(ad: KidozRewardedAd) {
            fullScreenAds.remove(adId)
            sendEvent(adId, "closed")
        }
    }

    private fun sendEvent(
        adId: Int,
        event: String,
        arguments: Map<String, Any?> = emptyMap(),
    ) {
        val payload = HashMap<String, Any?>(arguments).apply {
            put("adId", adId)
            put("event", event)
        }
        mainHandler.post { channel.invokeMethod("onAdEvent", payload) }
    }

    private companion object {
        const val CHANNEL_NAME = "kidoz_ads"
        const val BANNER_VIEW_TYPE = "kidoz_ads/banner"
    }
}
